### PR TITLE
GDB

### PR DESCRIPTION
**Pull request by tolusha**
Debugging CPP applications

### CODE REVIEW

File: plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerServiceClientImpl.java [hunk 1]
```diff
@@ -66,9 +67,14 @@ protected GdbDebuggerServiceClientImpl(@Named("cheExtensionPath") String extPath
     @Override
     public Promise<DebuggerInfo> connect(@NotNull Map<String, String> connectionProperties) {
         final String requestUrl = baseUrl + "/connect";
-        final String params = "?host=" + connectionProperties.get(HOST.toString())
+        String params = "?host=" + connectionProperties.get(HOST.toString())
                               + "&port=" + connectionProperties.get(PORT.toString())
-                              + "&file=" + connectionProperties.get(FILE.toString());
+                              + "&file=" + connectionProperties.get(BINARY.toString());
+
+        if (connectionProperties.containsKey(SOURCES.toString())) {
```
Comment (sleshchenko) [inline plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerServiceClientImpl.java]: Think about replacing these lines with 

```
String sources = connectionProperties.containsKey(SOURCES.toString());
       if (sources != null) {
            params += "&sources=" + sources;
        }
```

File: plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerServiceClientImpl.java [hunk 2]
```diff
@@ -66,9 +67,14 @@ protected GdbDebuggerServiceClientImpl(@Named("cheExtensionPath") String extPath
...
+            params += "&sources" + connectionProperties.get(SOURCES.toString());
```
Comment (sleshchenko) [inline plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerServiceClientImpl.java]: params += "&sources__**_=**___" + connectionProperties.get(SOURCES.toString()); ?

File: plugins/plugin-debugger/che-plugin-debugger-ide/src/main/java/org/eclipse/che/ide/ext/debugger/client/debug/AbstractDebugger.java [hunk 3]
```diff
@@ -675,5 +679,7 @@ protected void restoreDebuggerInfo() {
      */
     abstract protected List<String> resolveFilePathByLocation(@NotNull Location location);
 
+    abstract protected String pathToFqn(VirtualFile file);
```
Comment (sleshchenko) [inline plugins/plugin-debugger/che-plugin-debugger-ide/src/main/java/org/eclipse/che/ide/ext/debugger/client/debug/AbstractDebugger.java:683]: Add @Nullable here

**Comment by sleshchenko** [comment]
OK for me

File: plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerClient.java [hunk 4]
```diff
@@ -63,24 +69,37 @@ public GdbDebuggerClient(GdbDebuggerServiceClientImpl service,
...
     @Override
     protected List<String> resolveFilePathByLocation(@NotNull Location location) {
-        Collections.singleton(location.getClassName());
-        return Collections.singletonList(location.getClassName());
+        CurrentProject currentProject = appContext.getCurrentProject();
+        if (currentProject == null) {
+            return Collections.singletonList(location.getClassName());
+        }
+
+        String projectPath = appContext.getCurrentProject().getProjectConfig().getPath();
```
Comment (azatsarynnyy) [inline plugins/plugin-gdb/che-plugin-gdb-ide/src/main/java/org/eclipse/che/ide/gdb/client/GdbDebuggerClient.java]: you can reuse variable `currentProject` here

**Comment by vparfonov** [comment]
ok

File: plugins/plugin-gdb/che-plugin-gdb-server/src/test/java/org/eclipse/che/ide/gdb/server/parser/GdbInfoProgramTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    public void testProgramIsStopped() throws Exception {
+        GdbOutput gdbOutput = GdbOutput.of("Debugging a target over a serial line.\n" +
+                                           "Program stopped at 0x7ffff7ddb2d0.\n" +
+                                           "It stopped with signal SIGTRAP, Trace/breakpoint trap.\n");
+
+        GdbInfoProgram gdbInfoProgram = GdbInfoProgram.parse(gdbOutput);
+        assertEquals(gdbInfoProgram.getStoppedAddress(), "0x7ffff7ddb2d0");
+    }
+}
```
Comment (skabashnyuk) [inline plugins/plugin-gdb/che-plugin-gdb-server/src/test/java/org/eclipse/che/ide/gdb/server/parser/GdbInfoProgramTest.java]: new line

**Pull request merged**